iree/compiler/Dialect/Util/Transforms/GreedyHoistIntoGlobals.cpp

```diff
@@ -0,0 +1,308 @@
+          //   - It is non-const-expr
+          //   - It is const-expr but is a non hoistable leaf per our policy
+          //     (must match above check during forward iteration).
+          //   - It is legal to hoist this operand
+          if (targetInfo && targetInfo->isConstExpr() &&
+              isHoistableLeaf(targetInfo)) {
+            continue;
+          }
+          if (!canHoistOperand(&operand)) continue;
+
+          // Bingo.
+          LLVM_DEBUG(dbgs() << "HOIST CONST-EXPR:\n");
+          LLVM_DEBUG(dbgs() << "  : Operand #" << operand.getOperandNumber()
+                            << " of " << *operand.getOwner() << "\n");
+          LLVM_DEBUG(dbgs() << "  : From " << operand.get() << "\n\n");
+
+          hoistConstExpr(operand, hoistedMap, moduleSymbols, constExprs);
+        }
+      }
+
+      // // We only want to look at const-expr ops (non roots) since they may
+      // // have interesting escapes.
+      // if (!constExprs.isConstExprOperation(childOp)) {
+      //   return WalkResult::advance();
+      // }
+
+      // LLVM_DEBUG(dbgs() << "PROCESSING CONST-EXPR OP: " << *childOp << "\n");
+      // for (Value constExprResult : childOp->getResults()) {
+      //   SmallVector<OpOperand *> escapeOperands =
+      //       constExprs.getNonConstExprEscapes(constExprResult);
+      //   LLVM_DEBUG(dbgs() << "  : Escapes " << escapeOperands.size()
+      //                     << " to non-const-expr ops\n");
+      //   for (OpOperand *operand : escapeOperands) {
+      //     if (canHoistOperand(operand)) {
+      //       // Bingo.
+      //       Operation *targetOp = operand->getOwner();
+      //       LLVM_DEBUG(dbgs() << "HOIST CONST-EXPR:\n");
+      //       LLVM_DEBUG(dbgs() << "  : Operand #" <<
+      //       operand->getOperandNumber()
+      //                         << " of " << *targetOp << "\n");
+      //       LLVM_DEBUG(dbgs() << "  : From " << operand->get() << "\n\n");
+
+      //       hoistConstExpr(targetOp, operand, hoistedMap, moduleSymbols,
+      //                      constExprs);
+      //     } else {
+      //       LLVM_DEBUG(dbgs() << "CANNOT HOIST CONST-EXPR OPERAND #"
+      //                         << operand->getOperandNumber() << " of "
+      //                         << operand->getOwner() << "\n");
+      //     }
+      //   }
+      // }
+      return WalkResult::advance();
+    });
+
+    cleanupDeadOps(constExprs);
+  }
+
+  void hoistConstExpr(OpOperand &operand, HoistedValueMap &hoistedMap,
+                      SymbolTable &moduleSymbols,
+                      const ConstExprAnalysis &constExprs) {
+    Operation *targetOp = operand.getOwner();
+    Value originalValue = operand.get();
+    GlobalOp existingGlobal = hoistedMap.lookup(originalValue);
+
+    if (!existingGlobal) {
+      // No existing mapping.
+      Location loc = originalValue.getLoc();
+      OpBuilder builder = getModuleEndBuilder();
```

Comment:
we should insert these at the top or immediately prior to the root op first using them - this keeps the IR consistent with our other passes and makes it easier to ensure proper initialization order - by inserting at the end it ensures that all other initializers that may access these globals (indirectly or otherwise) have to be inserted after them (unless we write a step to try to sort them... but that sounds hard)